instant-xml-macros/src/de.rs

```diff
@@ -0,0 +1,289 @@
+
+                struct StructVisitor;
+                impl<'xml> Visitor<'xml> for StructVisitor {
+                    type Value = #ident;
+
+                    fn visit_struct<'a, D>(&self, deserializer: &mut D, attributes: Option<&std::collections::HashMap<String, String>>) -> Result<Self::Value, ::instant_xml::Error>
+                    where
+                        D: ::instant_xml::DeserializeXml<'xml> + ::instant_xml::AccessorXml<'xml>,
+                    {
+                        #declare_values
+                        println!("visit struct");
+                        while let Some(item) = &deserializer.peek_next_tag()? {
+                            match item {
+                                XmlRecord::Open(item) => {
+                                    match get_element(&item.key.as_ref()) {
+                                        #elem_type_match
+                                        __Elements::__Ignore => todo!(),
+                                    }
+
+                                    // Verify prefix
+                                    if let Some(prefix) = &item.prefix {
+                                        // Check if such prefix exist
+                                        if !deserializer.verify_prefix(&prefix) {
+                                            return Err(::instant_xml::Error::UnexpectedPrefix);
+                                        }
+                                        // TODO: Check if prefix is equel to declared prefix
+                                    }
+                                 }
+                                 XmlRecord::Close(tag) => {
+                                    println!("Close: {}", tag);
+                                    if tag == #name {
+                                        break;
+                                    }
+                                },
+                                XmlRecord::Element(_) => panic!("Unexpected element"),
+                            }
+                        }
+
+                        if let Some(attributes_map) = &attributes {
+                            for (key, value) in attributes_map.iter() {
+                                match get_attribute(&key) {
+                                    #attr_type_match
+                                    __Attributes::__Ignore => todo!(),
+                                }
+                            }
+                        }
+
+                        println!("return");
+                        Ok(Self::Value {
+                            #return_val
+                    })
+                    }
+                }
+
+                #namespaces_map;
+                deserializer.deserialize_struct(StructVisitor{}, #name, &namespaces_map)
+            }
+        ))
+        .into();
+
+        Deserializer {
+            input,
+            fn_deserialize,
+            fn_from_xml,
+        }
+    }
+
+    fn is_scalar(value: &str) -> bool {
```

Comment:
Should fix this like on the serializer side.

Reply:
There is different order here then in Serializer. 
This is producing something like:
```
        enum __Elements {
            __Value0,
            __Value2,
            __Ignore,
        }
        fn get_element(value: &str) -> __Elements {
            match value {
                "flag" => __Elements::__Value0,
                "Nested" => __Elements::__Value2,
                _ => __Elements::__Ignore,
            }
        }
```
This means that you are expecting the `Nested` tag that will be later deserialized to `Nested` struct. You can overide it when manually implementing the Deserialaizer. But to make it generic you need to assume somthing, or have something like `get_name()` which would return the default name, in Deserialize trait for each type. Assuming that non-scalars types by default have the struct name as tag name, looks clearer for me.

The other option would be the attribute like `rename` so you would not have to implement manually the whole thing, but still by default for non-scalars, i would keep the struct name as default.

Reply:
Having a hard-coded list of type names simply won't work (for example, `Option` might instead be spelled as `core::option::Option`), we need to defer that to a trait implementation.

Reply:
Good point, but we will still need this function `is_scalar`, as for scalar types we always want to use struct field name as a tag, not the default name from trait.

Reply:
Sure, we will need to interrogate the `FromXml` impl to see if it's a scalar.

Reply:
done, I added the function tag_name